### PR TITLE
Optimization of Sync Records: Implementing Pagination and Temporary Table

### PR DESCRIPTION
**Pull request by skrashevich**
This pull request introduces significant changes to optimize the process of syncing records, resulting in a speed optimization of up to 400 times. A temporary table is created for record deletion, and pagination is used to process recordings in chunks. This strategy is particularly beneficial for handling the deletion of recordings with missing files efficiently and swiftly

Speed Comparison:

Current:
```
Start sync recordings.
Deleting 32124 recordings with missing files
End sync recordings.
Old - Processed for 3046.11 seconds.
```
This PR:
```
Start sync recordings.
Deleting 31536 recordings with missing files
End sync recordings.
New - Processed for 7.99 seconds.
```

### CODE REVIEW

File: frigate/record/cleanup.py [hunk 1]
```diff
@@ -18,6 +18,10 @@
 logger = logging.getLogger(__name__)
 
 
+class RecordingsToDelete(Model):
+    id = CharField(null=False, primary_key=False, max_length=30)
```
Comment (NickM-27) [inline frigate/record/cleanup.py:22]: I think it would be good to store this in the models.py file so they are all kept in one place

**Review by NickM-27 — CHANGES_REQUESTED** [review]

**Comment by blakeblackshear** [comment]
What is the performance for the os.walk when there are tens of thousands of files in the directory? I think this was one of the problems we ran into.

**Comment by skrashevich** [comment]
> What is the performance for the os.walk when there are tens of thousands of files in the directory? I think this was one of the problems we ran into.

% python3 test.py
Folders and files created successfully.
Total - Processed 50001 files for 0.27 seconds.

```
import os
import datetime


def create_folders_with_files():
    # Create five folders
    for i in range(1, 6):
        folder_name = f"Folder_{i}"
        os.makedirs(folder_name, exist_ok=True)

        # Create 10,000 files within each folder
        for j in range(1, 10001):
            file_name = f"{folder_name}/File_{j}.txt"
            with open(file_name, "w") as file:
                file.write("This is file content.")

    print("Folders and files created successfully.")


# Call the function to create the folders and files
create_folders_with_files()

start_time = datetime.datetime.now().timestamp()
# get all recordings files on disk and put them in a set
files_on_disk = {
    os.path.join(root, file) for root, _, files in os.walk(".") for file in files
}
duration = datetime.datetime.now().timestamp() - start_time
print(f"Total - Processed {len(files_on_disk)} files for {duration:.2f} seconds.")

```

**Comment by blakeblackshear** [comment]
We should think carefully about when this runs. If a user temporarily mapped their recordings incorrectly or their NAS disconnected, this would drop all their recordings on restart.

**Comment by skrashevich** [comment]
> We should think carefully about when this runs. If a user temporarily mapped their recordings incorrectly or their NAS disconnected, this would drop all their recordings on restart.

make sense. I'll think about mark event in database as unreachable instead of delete

**Comment by blakeblackshear** [comment]
I think it would be best to use a simple heuristic instead. What if we try and detect when >50% of the recordings would be deleted and stop the automatic sync. We could also add a button on the Storage page that triggers a sync in "force" mode which can override the fail safe.

We could also consider creating a hidden file in the recordings directory that we could check for on startup. If it's not there and we have recordings in the database, then we can assume there is an issue with the mounted storage.

File: frigate/record/cleanup.py [hunk 2]
```diff
@@ -215,34 +215,62 @@ def sync_recordings(self) -> None:
...
+                    recordings_to_delete.add(recording.id)
+
+        # convert back to list of dictionaries for insertion
+        recordings_to_delete = [
+            {"id": recording_id} for recording_id in recordings_to_delete
+        ]
+
+        if len(recordings_to_delete) / recordings.count() > 0.5:
+            logger.debug(
+                f"Deleting {(len(recordings_to_delete) / recordings.count()):2f}% of recordings looks like as bug. Do nothing"
```
Comment (NickM-27) [inline frigate/record/cleanup.py:244]:
```suggestion
                f"Deleting {(len(recordings_to_delete) / recordings.count()):2f}% of recordings could be due to configuration error. Aborting..."
```

File: frigate/record/cleanup.py [hunk 3]
```diff
@@ -215,34 +215,62 @@ def sync_recordings(self) -> None:
...
+                if recording.path not in files_on_disk:
+                    recordings_to_delete.add(recording.id)
+
+        # convert back to list of dictionaries for insertion
+        recordings_to_delete = [
+            {"id": recording_id} for recording_id in recordings_to_delete
+        ]
+
+        if len(recordings_to_delete) / recordings.count() > 0.5:
+            logger.debug(
```
Comment (NickM-27) [inline frigate/record/cleanup.py:243]: Maybe we should make this log a warning to alert the user about potential misconfiguration for example in the event that their NAS mount failed?

Comment (skrashevich) [reply frigate/record/cleanup.py:243]: I believe that notifying user about the disconnected NAS should be placed elsewhere, definitely not in a internal utility function. 🙂

**Review by NickM-27 — APPROVED** [review]

**Comment by blakeblackshear** [comment]
Looks like there are a few conflicts here, but this is good to merge when those are resolved.

**Pull request merged**